guacamole-docker/bin/start.sh

```diff
@@ -30,7 +30,7 @@
 
 GUACAMOLE_HOME_TEMPLATE="$GUACAMOLE_HOME"
 
-GUACAMOLE_HOME="$HOME/.guacamole"
+GUACAMOLE_HOME="/tmp/guacamole"
```

Comment:
I really don't like this part of it - making the `GUACAMOLE_HOME` into a temp directory doesn't seem wise to me.  Maybe I'm thinking of this too much from a non-docker perspective, but what if the user decides to try to map through `/etc/guacamole` to a specific path?

I don't know if there's any other solution for this in the Docker Tomcat world, but this approach just makes me twitch :-).

Reply:
First, this line is due to the fact the user chosen to run the container may not have a home directory into the container, which would then not start.
Note however that the `guacamole` directory, inside the `tmp` directory, is only reachable by the user running the container, as shown below, from a container running this PR :
```
$ ls -l /tmp/
total 12
drwx------ 5 1024 1024 4096 Mar  4 13:02 guacamole
```

> but what if the user decides to try to map through `/etc/guacamole` to a specific path?

I also do this in my `docker-compose.yml` file, not an issue at all :
```
services:
  guacamole:
    image: guacamole/guacamole:1.1.0
    volumes:
      - ./etc:/etc/guacamole:ro
```

Reply:
My issue with the directory being placed in /tmp is not the accessibility of the directory to other users/processes, it's the fact that a configuration directory (or webapp deployment directory) is in /tmp - this doesn't make sense to me.  Configuration directories should be in configuration directory locations - not in any place that happens to work because we have write access to it :-).

I tend to agree with @manolan1, here - Guacamole can't be the only piece of software fighting this issue.  Is what you've done here a commonly-implemented solution among other Tomcat + Docker web applications that run as non-root users?

I'm willing to defer on this one if @mike-jumper or any of the other project members think we should just move forward, but this still just feels a little off to me.

Reply:
> Guacamole can't be the only piece of software fighting this issue. Is what you've done here a commonly-implemented solution among other Tomcat + Docker web applications that run as non-root users?

It is what Tomcat documentation recommends, it's the goal of `CATALINA_BASE` env var we then use in this PR.
IMO, what is done in the Tomcat Docker image, 777-chmoding application / configuration / work / tmp... default directories is not the best way to go.
Do you, on a standard installation, as non-root user, have write access to these directories ? No :-)
However you're free to set your own `CATALINA_BASE` directory, in your home directory for example, and start a new Tomcat instance over it.
As we do here in this PR.

> Configuration directories should be in configuration directory locations - not in any place that happens to work because we have write access to it :-)

We could `chmod 777 /home` at image build time (in `Dockerfile`), and then use `GUACAMOLE_HOME="/home/guacamole"` instead of `GUACAMOLE_HOME="/tmp/guacamole"`, but the whole idea would remain the same.
Keep in mind that whatever is done when Docker container starts will be lost when it will stop, and will have to be done again (this is what `start.sh` does).
So, configuration which is done when container starts, will be lost when it will stop, wherever it stands, in `/tmp`, `/etc`, `/home` etc...

I've just verified, there's nothing in `/home` in the Tomcat Docker image.
We could then safely `chmod 777` it and move `GUACAMOLE_HOME` into it, if you're more confortable with this.

I think this PR is the proper way to go to address this (non-)root security issue.
Of course I run it since the beginning without any issue.

Thank you again :+1:

Reply:
> Guacamole can't be the only piece of software fighting this issue. Is what you've done here a commonly-implemented solution among other Tomcat + Docker web applications that run as non-root users?

It is what Tomcat documentation recommends, it's the goal of `CATALINA_BASE` env var we then use in this PR.
IMO, what is done in the Tomcat Docker image, 777-chmoding application / configuration / work / tmp... default directories is not the best way to go.
Do you, on a standard installation, as non-root user, have write access to these directories ? No :-)
However you're free to set your own `CATALINA_BASE` directory, in your home directory for example, and start a new Tomcat instance over it.
As we do here in this PR.

> Configuration directories should be in configuration directory locations - not in any place that happens to work because we have write access to it :-)

We could `chmod 777 /home` at image build time (in `Dockerfile`), and then use `GUACAMOLE_HOME="/home/guacamole"` instead of `GUACAMOLE_HOME="/tmp/guacamole"`, but the whole idea would remain the same.
Keep in mind that whatever is done when Docker container starts will be lost when it will stop, and will have to be done again (this is what `start.sh` does).
So, configuration which is done when container starts, will be lost when it will stop, wherever it stands, in `/tmp`, `/etc`, `/home` etc...

I've just verified, there's nothing in `/home` in the Tomcat Docker image.
We could then safely `chmod 777` it and move `GUACAMOLE_HOME` into it, if you're more confortable with this.

We could also come back to the original `GUACAMOLE_HOME="$HOME/.guacamole"`, thus getting rid of the hard-coded `/tmp` path.
Everything will run as today, from inside `/root` directory, for those running the container as `root`.
The others, switching to a different user, will have to set the `$HOME` environment variable to a writable directory.
Could then suit everyone's needs here.

I think this PR is the proper way to go to address this (non-)root security issue.
Of course I run it since the beginning without any issue.

Thank you again :+1:

Reply:
So @necouchman ? :)
As per my previous answers, do you think we could go like this (or with one of the 2 modifications proposed above, if you prefer) ?
Many thanks 👍